### ISSUE TITLE
Do I need to configure anything in AndroidManifest.xml？

### ISSUE THREAD

**Issue by Rclsss** [alexrintt/shared-storage#117]
I want to use the `openDocumentTree` method to authorize the `Android/data/somedirectory/` Directory
Then use the `getDocumentContent `method to read or write sub files in the directory, so I don't need to authorize the files I want to read and write again through the `openDocument `method.

Because my current situation is that if I edit the file using another app (an app of the file management editing type), the sub file permissions that I authorize using the `openDocument `method within my application will become invalid, and then I need to call the `openDocument `method again to let the user select the file to regain read and write permissions.

So I want to make sure that my application won't invalidate my file permissions when other applications edit the file. Do I need to configure something in AndroidManifest.xml? How can I achieve this

**Comment by alexrintt** [comment]
> So I want to make sure that my application won't invalidate my file permissions when other applications edit the file.

This is the default behavior, once the user grant permission over a Uri, you have permission over all subfiles (recursively), so even if other app edit a file that is in under your granted tree, you still have access to it, also applies for created files.

I need some code and details to know what is exactly happening in your case, can you show how you are using and what is your target Android version?

---

A side-note: `Android/data` folder is not permitted to be selected anymore (even if the user wants!) on higher Android versions (12+), so be aware of this limitation (same applies for `MANAGE_EXTERNAL_STORAGE` permission) see https://www.reddit.com/r/Android/comments/wru35i/clearing_up_confusion_about_how_to_access/.

**Comment by alexrintt** [comment]
TL;DR: Try declaring the tree owner of your document at `fUri`.

```diff
final fUri = Uri.parse(
- 'content://com.android.externalstorage.documents/document/primary%3AAndroid%2Fdata%2Fmark.via%2Ffiles%2Flogs%2Fapp.ini'
+ 'content://com.android.externalstorage.documents/tree/primary%3AAndroid%2Fdata%2Fmark.via/document/primary%3AAndroid%2Fdata%2Fmark.via%2Ffiles%2Flogs%2Fapp.ini'
);
```

---

What I could understand from your functions is that:

- When you manually select the file, your code works because the returned Uri is containing the tree path plus the document uri, as it should be.
- When you select the tree, you are trying to read the document path without the tree path, AFAIK this leads to `SecurityException`.

So here is an example that creates your `dUri` and `fUri` correctly and can possible fix your issue:

```dart
void main() {
  // Granted tree uri:
  final dUri = Uri(
    scheme: 'content',
    host: 'com.android.externalstorage.documents',
    path: '/tree/primary%3AAndroid%2Fdata%2Fmark.via',
  );

  // content://com.android.externalstorage.documents/tree/primary%3AAndroid%2Fdata%2Fmark.via
  print(dUri);

  // Sub-file uri:
  final fUri = dUri.replace(
    path: dUri.path +
        '/document/primary%3AAndroid%2Fdata%2Fmark.via%2Ffiles%2Flogs%2Fapp.ini',
  );

  // content://com.android.externalstorage.documents/tree/primary%3AAndroid%2Fdata%2Fmark.via/document/primary%3AAndroid%2Fdata%2Fmark.via%2Ffiles%2Flogs%2Fapp.ini
  print(fUri);
}
```

**Comment by alexrintt** [comment]
TD;DR: Follow these steps:

1. Use the following template:
```
content://com.android.externalstorage.documents/tree/{encodedTreeUri}/document/{encodedSubfileUri}
```
3. Replace `encodedTreeUri` **with the URL encoded version** of `primary:{path-to-your-desired-storage-tree}`, in your case: `primary:Android/data/mark.via` (The path must not start with a `/`).
4. Replace `encodedSubfileUri` **with the URL encoded version** of `primary:{path-to-your-desired-storage-file}`, in your case: `primary:Android/data/mark.via/files/logs/app.ini` (The path must not start with a `/`).
5. The final result is the subfile (aka child file) URI.

---

Dart implementation:

```dart
final Uri dUri = createTreeUri('Android/data/mark.via');
final Uri fUri =
    createDocumentUri(dUri, 'Android/data/mark.via/files/logs/app.ini');

Uri createTreeUri(
  String treePath, {
  String host = 'com.android.externalstorage.documents',
  String scheme = 'content',
  String storage = 'primary',
}) {
  assert(
    !treePath.startsWith('/'),
    'Please, remove `/` from the start of your [treePath].',
  );

  final Uri treeUri = Uri(
    host: host,
    scheme: scheme,
    path: '/tree/${Uri.encodeComponent('$storage:$treePath')}',
  );

  return treeUri;
}

Uri createDocumentUri(
  Uri treeUri,
  String documentPath, {
  String storage = 'primary',
}) {
  assert(
    !documentPath.startsWith('/'),
    'Please, remove `/` from the start of your [documentPath].',
  );

  final Uri documentUri = treeUri.replace(
    path:
        '${treeUri.path}/document/${Uri.encodeComponent('$storage:$documentPath')}',
  );

  return documentUri;
}
```

I must warn that this is specific for your use-case: generating a pre-defined Uri from a pre-defined path that will be in the primary storage of the device using the default android storage provider. This will not work for all use-cases, like generating a URI for a cloud storage provider or even a custom storage provider. This also may not work for some devices.

**Comment by alexrintt** [comment]
Do not re-use closed issues, open a new one.
